slack connector: implemented signing secret and updated docs (#7186)
RasaHQ/rasa: docs/docs/connectors/slack.mdx

title: Slack
description: Build a Rasa Chat Bot on Slack
---
import createAppImg from './img/slack-create-app.png';
import eventsImg from './img/slack-events.png';
import installAppImg from './img/slack-install-app.png';
import interactivityImg from './img/slack-interactivity.png';
import requestUrlImg from './img/slack-request-url.png';
import scopesImg from './img/slack-scopes.png';
import secretImg from './img/slack-secret.png';

Connecting a bot to Slack requires you to configure it to send messages
(using API credentials) and to receive messages (using a webhook).

## Sending Messages

> Create a new file `credentials.yml` in the root folder of your rasa open source

Reviewer: ```suggestion
Create a new file `credentials.yml` in the root folder of your Rasa Open Source
```

   - `im:history`,
   - `mpim:history` and
   - `reactions:write`.

   <Image img={scopesImg} caption="Set up Slack Permissions" alt="Set up Slack Permissions Screenshot" max-width="500px" />

   In Slacks API documentation you can find a
   [list and explanation of all available scopes](https://api.slack.com/scopes).

3. On the **OAuth & Permissions** page, click **Install App to Workspace** to add
   the bot to your workspace.

   <Image img={installAppImg} caption="Install Slack App" alt="Install Slack App Screenshot" max-width="500px"/>

   Once added, Slack will show you a **Bot User OAuth Access Token** which you'll
   need to add to your `credentials.yml` as the value for `slack_token`:

   ```yaml-rasa title="credentials.yml" line={3}
   slack:
     slack_channel: "your-channel" # choose a channel for your bot
     slack_token: "xoxb-XXX"       # token obtained in the next step
     slack_signing_secret: "YYY"   # secret obtained in the next step
   ```

   The token should start with `xoxb`.

4. Head over to **Basic Information** to gather the **Signing Secret**.

   <Image img={secretImg} caption="Signing Secret" alt="Signing Secret Screenshot" max-width="500px"/>

>    Copy the signing secret into your `configuration.yml` as the value for `slack_signing_secret`:

Reviewer: ```suggestion
   Copy the signing secret into your `credentials.yml` as the value for `slack_signing_secret`:
```

   ```yaml-rasa title="credentials.yml" line={4}
   slack:
     slack_channel: "your-channel" # choose a channel for your bot
     slack_token: "xoxb-XXX"       # token obtained in the next step
     slack_signing_secret: "YYY"   # secret obtained in the next step
   ```

This setup will allow your bot to send messages. Now let's head over to the setup
for receiving and reacting to messages.

## Receiving Messages

> Make sure you've gone through [Sending Messages](./slack.mdx#sending-messages) first
> and you have a ready to go `credentials.yml`.

Reviewer: ```suggestion
Before continuing, make sure you have configured a Slack App for [Sending Messages](./slack.mdx#sending-messages) and have added Slack credentials to your `credentials.yml` file.
```

To receive messages, you will need a publicly available URL for Slack to reach
your bot and tell you about new messages. If you are running locally, you can

> [test channels using ngrok](../messaging-and-voice-channels#testing-channels-on-your-local-machine)

Reviewer: ```suggestion
[test channels using ngrok](../messaging-and-voice-channels#testing-channels-on-your-local-machine).
```

   Start your bot e.g. using
   ```bash
   rasa run
   ```
   If you are running locally, make sure ngrok (or another tool to retrieve a public
   url) is running as well.

2. Configure the webhook by heading to **Event Subscriptions** and
   turning **Enable Events** on.

   As a request URL enter the public url of your bot and append `/webhooks/slack/webhook`, e.g.
   `https://<host>/webhooks/slack/webhook` replacing `<host>` with your URL. If you
   are using ngrok, your url should look like `https://92832de0.ngrok.io/webhooks/slack/webhook`.

   You won't be able to use a `localhost` url.

   <Image img={requestUrlImg} caption="Request URL" alt="Request URL Screenshot" max-width="500px"/>

3. As a last step, you'll need to **Subscribe to bot events** on the same page.
   You'll need to add the following events:
   - `message.channels`,
   - `message.groups`,
   - `message.im` and
   - `message.mpim`.

   <Image img={eventsImg} caption="Subscribe to Bot Events" alt="Subscribe to Bot Events Screenshot" max-width="500px"/>

   Make sure to hit **Save Changes** at the bottom of the page after you've added these events.

>    (If you missed to grant all needed permissions to your app while setting up

Reviewer: ```suggestion
   (If you didn't grant all required permissions to your app while setting up
```

   sending of messages, you'll be prompted to **reinstall your app** which you will
   need to do. Otherwise, Slack will confirm your change with a **Success!**)

> Your bot is now ready to go and will receive webhook notifications about new mewssages.

Author: ```suggestion
Your bot is now ready to go and will receive webhook notifications about new messages.
```

## Optional: Interactive Components
After you've completed [Sending Messages](./slack.mdx#sending-messages) and
[Receiving Messages](./slack.mdx#receiving-messages) your bot is ready to go. If you

> want to use Slacks "Interactive" elements (buttons or menus), you'll need to do

Reviewer: ```suggestion
want to use Slack's [interactive components](https://api.slack.com/block-kit/interactivity) (buttons or menus), you'll need to do
```

some additional configuration:

Open the **Interactivity & Shortcuts** page and toggle **Interactivity** to be on.
Afterwards, you'll need to enter the same url into the **Request URL** field which

> you used in Step 2. of Receiving Messages, e.g. `https://<host>/webhooks/slack/webhook`.

Reviewer: ```suggestion
you used in Step 2 of [Receiving Messages](./slack.mdx#receiving-messages), e.g. `https://<host>/webhooks/slack/webhook`.
```
